source/fab/steps/preprocess.py

```diff
@@ -114,28 +117,29 @@ def process_artefact(self, fpath):
         return output_fpath
 
 
-class FortranPreProcessor(PreProcessor):
-    """
-    By default, preprocesses all .F90 and .f90 files in the *all_source* artefact,
-    creating the *preprocessed_fortran* artefact.
-
-    """
-    DEFAULT_SOURCE = SuffixFilter('all_source', '.F90')
-    DEFAULT_OUTPUT_NAME = 'preprocessed_fortran'
-    DEFAULT_OUTPUT_SUFFIX = '.f90'
-    LABEL = 'preprocess fortran'
-
-
-class CPreProcessor(PreProcessor):
-    """
-    By default, preprocesses all .c files in the *all_source* artefact, creating the *preprocessed_c* artefact.
-
-    An example of providing a :class:`~fab.util.ArtefactsGetter` would be
-    when preprocessing files which have come from the C pragma injector,
-    which creates the *pragmad_c* artefact.
-
-    """
-    DEFAULT_SOURCE = SuffixFilter('all_source', '.c')
-    DEFAULT_OUTPUT_NAME = 'preprocessed_c'
-    DEFAULT_OUTPUT_SUFFIX = '.c'
-    LABEL = 'preprocess c'
+def fortran_preprocessor(preprocessor='fpp', source=None,
+                         output_collection='preprocessed_fortran', output_suffix='.f90',
+                         name='preprocess fortran', **pp_kwds):
+
+    return PreProcessor(
+        preprocessor=preprocessor,
+        source=source or SuffixFilter('all_source', '.F90'),
+        output_collection=output_collection,
+        output_suffix=output_suffix,
+        name=name,
+        **pp_kwds
+    )
```

Comment:
Does this change point to an even wider generalisation of "takes a file, returns a file"?

Reply:
I'm struggling to connect those two ideas. Perhaps that's something for a chat? We should certainly keep looking for design improvements as we progress, please keep these thoughts coming!